File: static/caption-word-estimate.js
// Splits a multi-word CaptionWord-shaped object into per-word estimated sub-ranges by
// character offset within its own [t_start, t_end]; mirrors app/caption_word_estimate.py exactly,
// including counting by Unicode code point (not UTF-16 code unit) so astral-plane characters
// (emoji, etc.) don't drift the split between the JS preview and the Python ASS export.
// Exposes Timeline.estimateWordTimings. Depends on window.Timeline (load after timeline.js).
(() => {
  // Code-point length, matching Python's len() on a str (which counts code points, not
  // UTF-16 code units like JS's native .length would for astral-plane characters).
  function codePointLength(s) {
    return [...s].length;
  }

  function estimateWordTimings(word) {
    const tokens = word.text.split(/\s+/).filter(Boolean);
    if (tokens.length === 0) return [];
    const normalized = tokens.join(" ");
    const totalLen = codePointLength(normalized);
    const duration = word.t_end - word.t_start;
    const result = [];
    let offset = 0;
    tokens.forEach((token, i) => {
      const tokenLen = codePointLength(token);
      const startFrac = offset / totalLen;
      const endFrac = (offset + tokenLen) / totalLen;
      result.push({
        id: `${word.id}-${i}`,
        text: token,
        t_start: word.t_start + startFrac * duration,
        t_end: word.t_start + endFrac * duration,
      });
      offset += tokenLen + 1;
    });
    return result;
  }

  Object.assign(window.Timeline, { estimateWordTimings });
})();

File: app/caption_word_estimate.py
# Splits a multi-word CaptionWord (e.g. "talks about this" authored as one subtitle chunk)
# into per-word estimated sub-ranges, interpolated by character offset within its own
# [t_start, t_end]; a single-word entry passes through unchanged.
# Exposes estimate_word_timings. Consumed by app.ass_render.group_words. Depends on app.models.
from app.models import CaptionWord

def estimate_word_timings(word: CaptionWord) -> list[CaptionWord]:
    tokens = word.text.split()
    if not tokens:
        return []
    normalized = " ".join(tokens)
    total_len = len(normalized)
    duration = word.t_end - word.t_start
    result = []
    offset = 0
    for i, token in enumerate(tokens):
        start_frac = offset / total_len
        end_frac = (offset + len(token)) / total_len
        result.append(CaptionWord(
            id=f"{word.id}-{i}",
            text=token,
            t_start=word.t_start + start_frac * duration,
            t_end=word.t_start + end_frac * duration,
        ))
        offset += len(token) + 1
    return result
